### PR TITLE
JIT: Make effect handling in lowering less conservative

### LINKED ISSUE #92699: JIT: Failed ccmp recognition
```csharp
public struct Point
{
    public long X, Y;

    public static unsafe bool Equals(in Point left, in Point right)
    {
        return (left.X == right.X) & (left.Y == right.Y);
    }
}
```

generates
```asm
G_M28859_IG01:  ;; offset=0x0000
            stp     fp, lr, [sp, #-0x10]!
            mov     fp, sp
						;; size=8 bbWeight=1 PerfScore 1.50
G_M28859_IG02:  ;; offset=0x0008
            ldr     x2, [x0]
            ldr     x3, [x1]
            cmp     x2, x3
            cset    x2, eq
            ldr     x0, [x0, #0x08]
            ldr     x1, [x1, #0x08]
            cmp     x0, x1
            cset    x0, eq
            and     w0, w2, w0
						;; size=36 bbWeight=1 PerfScore 14.50
G_M28859_IG03:  ;; offset=0x002C
            ldp     fp, lr, [sp], #0x10
            ret     lr
						;; size=8 bbWeight=1 PerfScore 2.00
```

I would expect to see `ccmp`.

### PR DESCRIPTION
**Pull request by jakobbotsch**
The interference checking in lowering bases some of its checks on GenTree::gtFlags. This is conservative since it includes effect flags of operands, which does not make much sense for LIR.

This PR replaces the relevant uses of gtFlags with a new GenTree::OperEffects() that computes the relevant effect flags for the node, excluding operands. We already know how to recompute effect flags other than GTF_GLOB_REF and GTF_ORDER_SIDEEFF. This PR adds functions for these as well (the GTF_GLOB_REF version
GenTree::OperRequiresGlobRefFlag is courtesy of @SingleAccretion).

For GTF_ORDER_SIDEEFF we add a GenTree::OperSupportsOrderingSideEffect which captures explicitly (and conservatively) the current cases where we are setting the flag, and only allows these cases to support the flag. Setting the flag for other cases may result in the flag being removed or ignored.

Fix #92699

Example:
```csharp
public struct Point
{
    public long X, Y;

    public static unsafe bool Equals(in Point left, in Point right)
    {
        return (left.X == right.X) & (left.Y == right.Y);
    }
}
```
Codegen before:
```asm
G_M28859_IG02:  ;; offset=0x0008
            ldr     x2, [x0]
            ldr     x3, [x1]
            cmp     x2, x3
            cset    x2, eq
            ldr     x0, [x0, #0x08]
            ldr     x1, [x1, #0x08]
            cmp     x0, x1
            cset    x0, eq
            and     w0, w2, w0
```

Codegen after:
```asm
G_M28859_IG02:  ;; offset=0x0008
            ldr     x2, [x0]
            ldr     x3, [x1]
            ldr     x0, [x0, #0x08]
            ldr     x1, [x1, #0x08]
            cmp     x2, x3
            ccmp    x0, x1, 0, eq
            cset    x0, eq
```

### CODE REVIEW

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @BruceForstall @kunalspathak 

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=421055&view=ms.vss-build-web.run-extensions-tab), mainly from new containment opportunities and `ccmp` opportunities like in the issue.
TP diffs are mostly positive, except for the native linux-x64 TP runs that seem to be somewhat worse affected, with a libraries.pmi outlier of up to +0.21%.

File: src/coreclr/jit/gentree.cpp [hunk 1]
```diff
@@ -7026,6 +7025,153 @@ bool GenTree::OperMayThrow(Compiler* comp)
...
+        case GT_XAND:
+        case GT_XCHG:
+        case GT_LOCKADD:
+        case GT_CMPXCHG:
+        case GT_MEMORYBARRIER:
+        case GT_KEEPALIVE:
+            return true;
+
+        case GT_CALL:
+            return AsCall()->HasSideEffects(comp, /* ignoreExceptions */ true);
```
Comment (SingleAccretion) [inline src/coreclr/jit/gentree.cpp:7073]: This was apparently in my second commit - `GT_ALLOCOBJ` should be here with `return AsAllocObj()->gtHelperHasSideEffects;`

File: src/coreclr/jit/gentree.cpp [hunk 2]
```diff
@@ -25345,6 +25491,15 @@ bool GenTreeHWIntrinsic::OperRequiresCallFlag() const
     return false;
 }
 
+//------------------------------------------------------------------------------
+// OperRequiresCallFlag : Check whether the operation requires GTF_GLOB_REF flag regardless
```
Comment (SingleAccretion) [inline src/coreclr/jit/gentree.cpp:25495]:
```suggestion
// OperRequiresGlobRefFlag : Check whether the operation requires GTF_GLOB_REF flag regardless
```
C&P.

File: src/coreclr/jit/gentree.cpp [hunk 3]
```diff
@@ -7026,6 +7025,153 @@ bool GenTree::OperMayThrow(Compiler* comp)
...
+    {
+        flags &= ~GTF_EXCEPT;
+    }
+
+    if (((flags & GTF_GLOB_REF) != 0) && !OperRequiresGlobRefFlag(comp))
+    {
+        flags &= ~GTF_GLOB_REF;
+    }
+
+    if ((flags & GTF_ORDER_SIDEEFF) != 0 && !OperSupportsOrderingSideEffect())
```
Comment (SingleAccretion) [inline src/coreclr/jit/gentree.cpp:7167]:
```suggestion
    if (((flags & GTF_ORDER_SIDEEFF) != 0) && !OperSupportsOrderingSideEffect())
```

File: src/coreclr/jit/gentree.cpp [hunk 4]
```diff
@@ -7026,6 +7025,153 @@ bool GenTree::OperMayThrow(Compiler* comp)
...
+// Return Value:
+//   True if the given operator supports GTF_ORDER_SIDEEFF.
+//
+// Remarks:
+//   A node will still have this flag set if an operand has it set, even if the
+//   parent does not support it. This situation indicates that reordering the
+//   parent may be ok as long as it does not break ordering dependencies of the
+//   operand.
+//
+bool GenTree::OperSupportsOrderingSideEffect() const
```
[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/jit/gentree.cpp:7100]: I can add assert in fgdiagnostic to ensure that `GTF_ORDER_SIDEEFF` implies that `OperSupportsOrderingSideEffect() || (any operand has the flag)`.

**Comment by jakobbotsch** [comment]
The failures are known according to build analysis. The Fuzzlyn failures look like #92537

**Review by BruceForstall — APPROVED** [review]

File: src/coreclr/jit/gentree.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/jit/gentree.cpp:7100]: Doing this requires making sure we propagate/unset these flags in all the same places as the other flags... That's probably a good thing to do down the road, but I think for now I will skip it and think about doing that separately.
Let me add something like `SetHasOrderingSideEffect()` with the assert in it for this PR.

[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/jit/gentree.cpp:7100]: Doing this did make me notice that I missed `GT_NULLCHECK` in `OperSupportsOrderingSideEffect` though, so definitely worth to have something.

File: src/coreclr/jit/gentree.h [hunk 5]
```diff
@@ -7177,6 +7185,12 @@ struct GenTreeIndir : public GenTreeOp
         return (gtFlags & GTF_IND_UNALIGNED) != 0;
     }
 
+    // True if this indirection is invariant.
+    bool IsInvariantLoad() const
+    {
+        return (gtFlags & GTF_IND_INVARIANT) != 0;
```
[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/jit/gentree.h]: Added it

File: src/coreclr/jit/gentree.cpp [hunk 6]
```diff
@@ -7026,6 +7025,153 @@ bool GenTree::OperMayThrow(Compiler* comp)
     return OperExceptions(comp) != ExceptionSetFlags::None;
 }
 
+//------------------------------------------------------------------------------
+// OperRequiresGlobRefFlag : Check whether the operation requires GTF_GLOB_REF
+//                           flag regardless of the children's flags.
+//
```
[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/jit/gentree.cpp:7031]: Expanded on this a bit in the remarks.

File: src/coreclr/jit/gentree.cpp [hunk 7]
```diff
@@ -7026,6 +7025,153 @@ bool GenTree::OperMayThrow(Compiler* comp)
...
+        case GT_XORR:
+        case GT_XAND:
+        case GT_XCHG:
+        case GT_LOCKADD:
+        case GT_CMPXCHG:
+        case GT_MEMORYBARRIER:
+        case GT_CATCH_ARG:
+#if defined(FEATURE_HW_INTRINSICS)
+        case GT_HWINTRINSIC:
+#endif
```
[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/jit/gentree.cpp]: Removed.

**Pull request merged**